tests/integ/test_udaf.py

```diff
@@ -0,0 +1,444 @@
+    Utils.check_answer(df.agg(VariantMax("v")), [Row('"v1"')])
+
+
+def test_array(session):
+    @udaf
+    class ArrayAgg:
+        def __init__(self) -> None:
+            self._agg_state = []
+
+        @property
+        def aggregate_state(self) -> List[Any]:
+            return self._agg_state
+
+        def accumulate(
+            self,
+            list_: List[str],
+            list_of_variants: List[Variant],
+        ) -> None:
+            self._agg_state += list_ + list_of_variants
+
+        def merge(self, other_state: List[Any]) -> None:
+            self._agg_state += other_state
+
+        def finish(self) -> List[str]:
+            return self._agg_state
+
+    df = session.sql(
+        "select array_construct('a1', 'a1') as a1, array_construct(to_variant('a2'), to_variant('a2')) as a2"
+    )
+    Utils.check_answer(
+        df.agg(ArrayAgg("a1", "a2")),
+        [Row('[\n  "a1",\n  "a1",\n  "a2",\n  "a2"\n]')],
+    )
+
+
+def test_object(session):
+    @udaf
+    class ObjectAgg:
+        def __init__(self) -> None:
+            self._agg_state = {}
+
+        @property
+        def aggregate_state(self) -> Dict[str, str]:
+            return self._agg_state
+
+        def accumulate(
+            self,
+            dict_: Dict[str, str],
+            dict_str_to_variant: Dict[str, Variant],
+        ) -> None:
+            self._agg_state.update(dict_)
+            self._agg_state.update(dict_str_to_variant)
+
+        def merge(self, other_state: Dict[str, str]) -> None:
+            self._agg_state.update(other_state)
+
+        def finish(self) -> Dict[str, str]:
+            return self._agg_state
+
+    df = session.sql(
+        "select object_construct('o1', 'o1') as o1, object_construct('o2', to_variant('v2')) as o2"
+    )
+    Utils.check_answer(
+        df.agg(ObjectAgg("o1", "o2")),
+        [Row('{\n  "o1": "o1",\n  "o2": "v2"\n}')],
+    )
+
+
+def test_register_udaf_from_file_without_type_hints(session, resources_path):
+    test_files = TestFiles(resources_path)
+    sum_udaf = session.udaf.register_from_file(
+        test_files.test_udaf_py_file,
+        "MyUDAFWithoutTypeHints",
+        return_type=IntegerType(),
+        input_types=[IntegerType()],
+    )
+    df = session.create_dataframe([[1, 3], [1, 4], [2, 5], [2, 6]]).to_df("a", "b")
+    Utils.check_answer(df.agg(sum_udaf("a")), [Row(6)])
+    Utils.check_answer(df.group_by("a").agg(sum_udaf("b")), [Row(1, 7), Row(2, 11)])
+
+
+def test_register_udaf_from_file_with_type_hints(session, resources_path):
+    test_files = TestFiles(resources_path)
+    sum_udaf = session.udaf.register_from_file(
+        test_files.test_udaf_py_file,
+        "MyUDAFWithTypeHints",
+    )
+    df = session.create_dataframe([[1, 3], [1, 4], [2, 5], [2, 6]]).to_df("a", "b")
+    Utils.check_answer(df.agg(sum_udaf("a")), [Row(6)])
+    Utils.check_answer(df.group_by("a").agg(sum_udaf("b")), [Row(1, 7), Row(2, 11)])
+
+
+def test_udaf_negative(session):
```

Comment:
Just a thought: ideally all negative tests could be unit tests 🤔

Reply:
Makes sense. Let me try moving them to unit tests.